### PR TITLE
Fix Island connection retries

### LINKED ISSUE #3117: ETE tests fail
## Describe the bug
Docker ETE tests fail (authentication issue)
AppImage fails to exploit some hosts

## Tasks
- [x] Fix Docker (0d)
- [ ] Fix AppImage (0.25d)

### PR DESCRIPTION
**Pull request by shreyamalviya**
# What does this PR do?

Fixes #3117 

## PR Checklist
* [x] Have you added an explanation of what your changes do and why you'd like to include them?
* [ ] Is the TravisCI build passing?
* [ ] ~Was the CHANGELOG.md updated to reflect the changes?~
* [ ] ~Was the documentation framework updated to reflect the changes?~
* [x] Have you checked that you haven't introduced any duplicate code?

## Testing Checklist

* [x] Added relevant unit tests?
* [x] Do all unit tests pass?
* [ ] Do all end-to-end tests pass?
* [x] Any other testing performed?
    > Tested by running the Agent locally 
* [ ] ~If applicable, add screenshots or log transcripts of the feature working~

### CODE REVIEW

File: monkey/infection_monkey/island_api_client/configuration_validator_decorator.py [hunk 1]
```diff
@@ -25,8 +25,8 @@ class ConfigurationValidatorDecorator(IIslandAPIClient):
     def __init__(self, island_api_client: IIslandAPIClient):
         self._island_api_client = island_api_client
 
-    def connect(self, island_server: SocketAddress):
-        return self._island_api_client.connect(island_server)
+    def login(self, otp: str):
```
Comment (VakarisZ) [inline monkey/infection_monkey/island_api_client/configuration_validator_decorator.py:28]: Off-topic, but isn't there something we could do to not have to thin-wrap all of the IIslandAPIClient methods? Most of them are irrelevant for this decorator... Maybe we could just override the methods in the constructor and return an instance of `IIslandAPIClient`? I remember we looked into it but maybe 3.11 has something better?

File: monkey/infection_monkey/island_api_client/i_island_api_client.py [hunk 2]
```diff
@@ -15,19 +15,19 @@ class IIslandAPIClient(ABC):
...
-        :param island_server: The socket address of the API
-        :raises IslandAPIConnectionError: If the client cannot successfully connect to the island
-        :raises IslandAPIRequestError: If an error occurs while attempting to connect to the
-                                       island due to an issue in the request sent from the client
-        :raises IslandAPIRequestFailedError: If an error occurs while attempting to connect to the
-                                             island due to an error on the server
-        :raises IslandAPITimeoutError: If a timeout occurs while attempting to connect to the island
-        :raises IslandAPIError: If an unexpected error occurs while attempting to connect to the
-                                island
+        :param otp: A one-time password used to authenticate with the Island API
```
Comment (VakarisZ) [inline monkey/infection_monkey/island_api_client/i_island_api_client.py:22]: NIT: authenticate **to** the Island API?

File: monkey/infection_monkey/island_api_client/i_island_api_client.py [hunk 3]
```diff
@@ -15,19 +15,19 @@ class IIslandAPIClient(ABC):
...
+        :raises IslandAPIConnectionError: If the client cannot successfully connect to the Island
```
Comment (VakarisZ) [inline monkey/infection_monkey/island_api_client/i_island_api_client.py:23]: All of these comments are missing AuthenticationError. I'm not sure this is the PR to fix it though

File: monkey/infection_monkey/network/relay/utils.py [hunk 4]
```diff
@@ -50,36 +42,27 @@ def find_available_island_apis(
...
     with suppress(StopIteration):
         server = next(servers)
-        server_results[server] = _check_if_island_server(server, island_api_client_factory)
+        server_results[server] = _check_if_island_server(server)
 
 
-def _check_if_island_server(
-    server: SocketAddress, island_api_client_factory: AbstractIslandAPIClientFactory
-) -> Optional[IIslandAPIClient]:
+def _check_if_island_server(server: SocketAddress) -> bool:
```
Comment (VakarisZ) [inline monkey/infection_monkey/network/relay/utils.py:51]: NIT: `_is_island_server` IMO better hints that it returns a bool and is a shorter name

**Review by VakarisZ — COMMENTED** [review]
Supreme work. My only question is why `IslandHTTPClient.login` is not part of the constructor?

File: monkey/infection_monkey/island_api_client/configuration_validator_decorator.py (hunk 1 above)
Comment (mssalvatore) [reply monkey/infection_monkey/island_api_client/configuration_validator_decorator.py:28]: We can investigate if 3.11 has something new we can use. I think we looked into this with 3.7 and hit walls.

File: monkey/infection_monkey/island_api_client/i_island_api_client.py (hunk 3 above)
Comment (mssalvatore) [reply monkey/infection_monkey/island_api_client/i_island_api_client.py:25]: It's not really for this PR, but it's a simple fix so I just did it.

File: monkey/infection_monkey/island_api_client/http_island_api_client.py [hunk 5]
```diff
@@ -50,28 +50,17 @@ def __init__(
...
-            self._http_client.server_url = f"https://{island_server}/api/"
-            self._http_client.get(params={"action": "is-up"})
-        except Exception as err:
-            self._http_client.server_url = None
-            raise err
-
-        auth_token = self._get_authentication_token()
-        self._http_client.additional_headers = {HTTPIslandAPIClient.TOKEN_HEADER_KEY: auth_token}
+    @handle_response_parsing_errors
+    def login(self, otp: str):
```
[the remark this answers is not in the document]
Comment (mssalvatore) [reply monkey/infection_monkey/island_api_client/http_island_api_client.py:58]: I've been bitten before when I've tried to make login happen in the constructor. So far, it's always ended up refactored this way.

File: monkey/infection_monkey/network/relay/utils.py (hunk 4 above)
Comment (mssalvatore) [reply monkey/infection_monkey/network/relay/utils.py]: Renamed to `_server_is_island()`. See https://guardicore.atlassian.net/wiki/spaces/RES/pages/2800353435/Infection+Monkey+Development+Guidelines+DRAFT#Functions/Methods-returning-booleans

**Pull request merged**